lib/CL/pocl_img_buf_cpy.c

```diff
@@ -0,0 +1,199 @@
+    POCL_RETURN_ERROR_ON((src->image_channel_order != dst->image_channel_order),
+      CL_IMAGE_FORMAT_MISMATCH, "src and dst have different image channel order\n");
+    POCL_RETURN_ERROR_ON((src->image_channel_data_type != dst->image_channel_data_type),
+      CL_IMAGE_FORMAT_MISMATCH, "src and dst have different image channel data type\n");
+    POCL_RETURN_ERROR_ON(((dst->type == CL_MEM_OBJECT_IMAGE2D ||
+          src->type == CL_MEM_OBJECT_IMAGE2D) &&  region[2] != 1),
+      CL_INVALID_VALUE, "for any 2D image copy, region[2] must be 1\n");
+  }
+
+
+  /* Images need to recompute the regions in bytes before copying */
+  size_t mod_region[3], mod_src_origin[3], mod_dst_origin[3];
+  memcpy(mod_region, region, 3*sizeof(size_t));
+  memcpy(mod_src_origin, src_origin, 3*sizeof(size_t));
+  memcpy(mod_dst_origin, dst_origin, 3*sizeof(size_t));
+
+  if (src_is_image && dst_is_image) {
+    mod_region[0] *= src->image_elem_size * src->image_channels;
+    mod_src_origin[0] *= src->image_elem_size * src->image_channels;
+    mod_dst_origin[0] *= dst->image_elem_size * dst->image_channels;
+  }
+
+  if (src_is_image) {
+    src_row_pitch = src->image_row_pitch;
+    src_slice_pitch = src->image_slice_pitch;
+  }
+
+  if (dst_is_image) {
+    dst_row_pitch = dst->image_row_pitch;
+    dst_slice_pitch = dst->image_slice_pitch;
+  }
+
+  POCL_RETURN_ERROR_ON(((command_queue->context != src->context) ||
+      (command_queue->context != dst->context)), CL_INVALID_CONTEXT,
+      "src, dst and command_queue are not from the same context\n");
+
+  POCL_RETURN_ERROR_COND((event_wait_list == NULL && num_events_in_wait_list > 0),
+    CL_INVALID_EVENT_WAIT_LIST);
+
+  POCL_RETURN_ERROR_COND((event_wait_list != NULL && num_events_in_wait_list == 0),
+    CL_INVALID_EVENT_WAIT_LIST);
+
+  size_t region_bytes = mod_region[0] * mod_region[1] * mod_region[2];
+  POCL_RETURN_ERROR_ON((region_bytes <= 0), CL_INVALID_VALUE, "All items in region must be >0\n");
+
+  if (pocl_buffer_boundcheck_3d(src->size, mod_src_origin, mod_region, &src_row_pitch,
+      &src_slice_pitch, "src_") != CL_SUCCESS) return CL_INVALID_VALUE;
+
+  if (pocl_buffer_boundcheck_3d(dst->size, mod_dst_origin, mod_region, &dst_row_pitch,
+      &dst_slice_pitch, "dst_") != CL_SUCCESS) return CL_INVALID_VALUE;
+
+  if (src == dst) {
+    POCL_RETURN_ERROR_ON((src_slice_pitch != dst_slice_pitch),
+      CL_INVALID_VALUE, "src and dst are the same object,"
+      " but the given dst & src slice pitch differ\n")
+
+    POCL_RETURN_ERROR_ON((src_row_pitch != dst_row_pitch),
+      CL_INVALID_VALUE, "src and dst are the same object,"
+      " but the given dst & src row pitch differ\n")
+
+    POCL_RETURN_ERROR_ON((check_copy_overlap(mod_src_origin, mod_dst_origin, mod_region,
+      src_row_pitch, src_slice_pitch)), CL_MEM_COPY_OVERLAP, "src and "
+      "dst are the same object, and source and destination "
+      "regions overlap");
+
+  }
+
+  POCL_CHECK_DEV_IN_CMDQ
```

Comment:
Could this macro made such that it allows adding a semicolon at the end of the line. I have had problems with editors getting confused and breaking auto indent after seeing this type of statements.